test/test_maskedtensor.py

```diff
@@ -536,24 +697,154 @@ def test_masked_bmm(self):
             ]
         )
         attn_mask = attn_mask_bool.float().masked_fill_(attn_mask_bool, float("-inf"))
-        v = masked_bmm(x, x_mt.transpose(1, 2), attn_mask)
-        v.sum().backward()
 
-    def test_linear(self):
-        x = torch.arange(4 * 3 * 2).reshape(4, 3, 2)
-        w_x = torch.arange(10).reshape(5, 2) + x.amax()
-        linear = torch.nn.functional.linear
-        key_padding_mask = torch.tensor(
-            [
-                [False, False, False, True],
-                [False, True, True, True],
-                [False, True, False, True],
-            ]
-        )
-        x_mt = MaskedTensor(
-            x, ~(key_padding_mask.transpose(0, 1).unsqueeze(-1).expand_as(x))
-        )
+        mt_res = masked_bmm(x, x_mt.transpose(1, 2), attn_mask)
+        mt_res.sum().backward()  # make sure this runs
```

Comment:
I've been thinking about this some more, and I might hav some fundamental misunderstanding here, but I don't believe the gradients should match.. please correct me if I misunderstood something cc @cpuhrsch since you originally authored `MaskedBmm` so likely know it better than I:

Consider the `MaskedBmm` implementation:

https://github.com/pytorch/pytorch/blob/3b9486461eac463771e45d6ecc0b3459cb199562/torch/masked/maskedtensor/matmul.py#L15-L41

and the derivative from `derivatives.yaml`:

https://github.com/pytorch/pytorch/blob/33352336b443dbfce1394f6b4950e8f33eff2cef/tools/autograd/derivatives.yaml#L348-L351

Indeed, `q_grad = torch.bmm(grad, k_transpose)` and `k_grad = torch.bmm(q_transpose, grad)`

In this testing, we've verified that `mt_res` is equivalent to `t_res` -- that is, the forward passes are equivalent, because we've masked out `k` when we call `x_masked = x.masked_fill(~x_mt.get_mask(), 0).detach().clone().requires_grad_()`

However, grad is not masked out in the backwards pass of vanilla `torch.bmm`, whereas it is implicitly in line 38 of the `MaskedBmm` implementation. 

In addition, instead of `torch.bmm(grad_data, k_trans)` (line 35 above) it should probably be `torch.bmm(grad.to_tensor(0), k_trans` -- is that right?